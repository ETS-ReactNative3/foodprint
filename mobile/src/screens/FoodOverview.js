import React, { useState } from 'react';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  Text,
  SafeAreaView,
  Platform,
  PermissionsAndroid,
  Image,
  View,
  ScrollView,
} from 'react-native';
import ImagePicker from 'react-native-image-picker';
import { TouchableOpacity } from 'react-native-gesture-handler';
import Axios from 'axios';
import Config from 'react-native-config';

const postPictureUri = Config.SERVER_URL + 'picture';

const options = {
  storageOptions: {
    // skipBackup: true,
    // path: 'images',
  },
};

const FoodOverview = ({ navigation }) => {
  const [food, setFood] = useState([]);

  const getFilePermissions = async () => {
    console.log('Ensuring file permissions.');
    try {
      if (Platform.OS === 'android') {
        const granted = await PermissionsAndroid.request(
          PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
          {
            title: 'App needs file permissions to save your photos',
            message: 'App needs file permissions to save your food photos',
            buttonNegative: 'Cancel',
            buttonPositive: 'OK',
          },
        );
        if (granted === PermissionsAndroid.RESULTS.GRANTED) {
          console.log('File permissions allowed');
        } else {
          console.log('File permissions denied');
        }
      }
    } catch (err) {
      console.warn(err);
    }
  };

  const takePicture = async () => {
    await getFilePermissions();

    ImagePicker.launchCamera(options, async response => {
      if (response.didCancel) {
        return console.log('User cancelled image picker');
      }
      if (response.error) {
        return console.log('ImagePicker Error: ', response.error);
      }
      if (response.customButton) {
        return console.log(
          'User tapped custom button: ',
          response.customButton,
        );
      }
      try {
        const data = new FormData();
        data.append('picture', {
          uri: response.uri,
          type: 'image/jpeg',
          name: 'pic.jpg',
        });
        const config = {
          headers: { 'content-type': 'multipart/form-data' },
          accept: 'application/json',
        };
        console.log({ postPictureUri });
        const carbonFootprintResponse = await Axios.post(
          postPictureUri,
          data,
          config,
        );
        console.log({ carbonFootprintResponse });

        if (!carbonFootprintResponse.data.error) {
          const meal = {
            description: carbonFootprintResponse.data.description,
            score: carbonFootprintResponse.data.score,
            uri: 'data:image/jpeg;base64,' + response.data,
          };
          setFood([...food, meal]);
        } else {
          console.warn('No meal found, handle this case for user.');
        }
      } catch (err) {
        console.warn({ err });
      }
    });
  };

  return (
    <SafeAreaView style={{ width: '100%', height: '100%' }}>
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          alignItems: 'center',
          padding: 8,
        }}>
        <Text style={{ fontSize: 24 }}>Your food history</Text>
        <TouchableOpacity onPress={takePicture}>
          <MaterialCommunityIcons name="plus" color={'black'} size={50} />
        </TouchableOpacity>
      </View>
      <ScrollView>
        {food.map((meal, i) => {
          console.log({ meal });
          return (
            <View key={i}>
              <Image
                style={{ width: 200, height: 400 }}
                source={{ uri: meal.uri }}
              />
              <Text>{meal.description}</Text>
              <Text>{meal.score}</Text>
            </View>
          );
        })}
      </ScrollView>
    </SafeAreaView>
  );
};

export default FoodOverview;
