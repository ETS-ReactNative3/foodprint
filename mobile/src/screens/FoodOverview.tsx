import React from 'react';
import {
  Text,
  Button,
  SafeAreaView,
  Platform,
  PermissionsAndroid,
} from 'react-native';
import ImagePicker from 'react-native-image-picker';

interface Props {}

const options = {
  storageOptions: {
    // skipBackup: true,
    // path: 'images',
  },
};

const FoodOverview: React.FC<Props> = ({navigation}) => {
  const takePicture = () => {
    // ask for file perms
    if (Platform.OS === 'android') {
      PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
        {
          title: 'App needs file permissions to save your photos',
          message: 'App needs file permissions to save your food photos',
          buttonNegative: 'Cancel',
          buttonPositive: 'OK',
        },
      );
    }

    ImagePicker.launchCamera(options, async response => {
      if (response.didCancel) {
        console.log('User cancelled image picker');
      } else if (response.error) {
        console.log('ImagePicker Error: ', response.error);
      } else if (response.customButton) {
        console.log('User tapped custom button: ', response.customButton);
      } else {
        // You can also display the image using data:
        // const source = { uri: 'data:image/jpeg;base64,' + response.data };

        try {
          const endpoint = 'http://8aacbaaa.ngrok.io/';
          const data = new FormData();
          data.append('picture', {
            uri: response.uri,
            type: 'image/jpeg',
            name: 'pic.jpg',
          });
          const config = {
            method: 'post',
            // headers: {'content-type': 'multipart/form-data'},
            body: data,
          };
          await fetch(endpoint, config);
        } catch (err) {
          console.log(err);
        }
        // and update state on response
      }
    });
  };

  return (
    <SafeAreaView>
      <Text>Food overview</Text>
      <Button onPress={takePicture} title="Open camera" />
    </SafeAreaView>
  );
};

export default FoodOverview;
