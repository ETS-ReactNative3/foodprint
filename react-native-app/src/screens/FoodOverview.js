import React from 'react';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  Text,
  SafeAreaView,
  Image,
  View,
  ScrollView,
} from 'react-native';
import { StyleSheet } from 'react-native';
import { useLayoutEffect } from 'react';
import { TouchableOpacity } from 'react-native-gesture-handler';

const FoodOverview = ({ navigation }) => {
  <MaterialCommunityIcons name="camera" color={'grey'} size={35} />;

  useLayoutEffect(() => {
    const takePicture = async () => {
      navigation.navigate('Camera');
    };

    navigation.setOptions({
      headerRight: () => (
        <TouchableOpacity onPress={takePicture} >
          <MaterialCommunityIcons name="camera" color={'grey'} size={28} style={styles.cameraButton} />
        </TouchableOpacity >
      ),
    });
  }, [navigation]);

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <View style={styles.contentContainer}>
          <Text style={styles.title}>Welcome</Text>
          <Text style={styles.subtitle}>
            Scan a barcode or snap a picture.
          </Text>
          <Image
            source={{ uri: 'https://png2.cleanpng.com/sh/4618a1d69c3326bc75ab0d6a0d3cd256/L0KzQYm3VcIyN5NvjJH0aYP2gLBuTfVud6Vue9H3LXXwf7vwTgN1cZRwfeQ2ZnHmdbP2jBsucZ9oReV2aXzoiX68gsAzPWNnfKdrM0LmQXA9WMIyOGE7T6MAMka7R4mCUMk0P2Q7RuJ3Zx==/kisspng-emoticon-emoji-sticker-facebook-inc-smiley-5b0252bd5b32c1.6821006715268789093736.png' }}
            style={styles.image}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: { width: '100%', height: '100%' },
  cameraButton: { marginRight: 16 },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 8,
  },
  contentContainer: { justifyContent: 'center', alignItems: 'center', marginHorizontal: 40 },
  title: { fontSize: 36, marginVertical: 20, textAlign: 'center' },
  subtitle: { fontSize: 20, textAlign: 'center', marginTop: 20 },
  image: { width: 200, height: 200, marginTop: 10 },
});

export default FoodOverview;
