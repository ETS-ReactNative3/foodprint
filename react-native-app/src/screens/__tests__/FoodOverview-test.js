import React from 'react';

import renderer from 'react-test-renderer';
import FoodOverview from '../FoodOverview';

// Check current component against its snapshot
test('FoodOverview matches previous snapshot', () => {
  const mockedNavigation = {
    setOptions: () => { },
  };
  const tree = renderer.create(<FoodOverview navigation={mockedNavigation} />).toJSON();
  expect(tree).toMatchSnapshot();
});
