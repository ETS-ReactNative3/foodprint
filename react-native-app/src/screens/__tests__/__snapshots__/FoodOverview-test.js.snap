// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodOverview matches previous snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <ScrollView>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginHorizontal": 40,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 36,
            "marginVertical": 20,
            "textAlign": "center",
          }
        }
      >
        Welcome
      </Text>
      <Text
        style={
          Object {
            "fontSize": 20,
            "marginTop": 20,
            "textAlign": "center",
          }
        }
      >
        Scan a barcode or snap a picture.
      </Text>
      <Image
        source={
          Object {
            "uri": "https://png2.cleanpng.com/sh/4618a1d69c3326bc75ab0d6a0d3cd256/L0KzQYm3VcIyN5NvjJH0aYP2gLBuTfVud6Vue9H3LXXwf7vwTgN1cZRwfeQ2ZnHmdbP2jBsucZ9oReV2aXzoiX68gsAzPWNnfKdrM0LmQXA9WMIyOGE7T6MAMka7R4mCUMk0P2Q7RuJ3Zx==/kisspng-emoticon-emoji-sticker-facebook-inc-smiley-5b0252bd5b32c1.6821006715268789093736.png",
          }
        }
        style={
          Object {
            "height": 200,
            "marginTop": 10,
            "width": 200,
          }
        }
      />
    </View>
  </ScrollView>
</SafeAreaView>
`;
